### ISSUE TITLE
Parser Value Set in PSA

### ISSUE THREAD

**Issue by hanw** [p4lang/p4-spec#565]
I have tried to implement parser value set in a couple different ways. One of the implementations is in [PR1024](https://github.com/p4lang/p4c/pull/1024). A second implementation is in branch [pvs-is-member](https://github.com/p4lang/p4c/tree/temp/pvs-is-member).

Both implementations have their pros and cons. I will summarize below:
PR1024 proposed to follow the P4-14 syntax and use the value_set instance name as the label in the select case. 
```
/// proposal in PR1024 and code example
extern value_set {
    value_set(bit<8> size);
}

value_set(8w4) pvs0_0;
value_set(8w4) pvs1_0;
state parse_ethernet {
   packet.extract<ethernet_t>(hdr.ethernet);
   transition select(hdr.ethernet.etherType) {
       pvs0_0: accept;
       pvs1_0: parse_inner_ethernet;
       default: accept;
   }
}
```
The issue, as rightfully raised by Mihai, is that the implementation contains a hack in the typeChecker to accept the value_set instance name (pvs0_0) as a 'compile-time constant' in the parser select cases, [here](https://github.com/p4lang/p4c/pull/1024/files#diff-ac7ed0f3ecd1057692c467bd565056abR2750) and [here](https://github.com/p4lang/p4c/pull/1024/files#diff-42bfc104320176fa10208502f2b9c87fR47). 

However, the positive side of the proposal is that the implementation of bmv2 backend and other backend is straightforward and proposal captures the semantics of the value set, i.e., the values represented by the name of the value set **is** a constant (at runtime).

The second implementation, with the `is_member` proposal, does not require the hack in the type checker. However, the backend implementation is much more complicated. One possible compilation strategy is to 'lower' the `is_member` proposal to the extern definition in the first proposal and then generates BMV2 JSON from the first proposal.

To summarize the second approach:
```
Step 1. represent value_set membership with is_member()

ValueSet<bit<16>>(4) pvs;
state parse_ethernet {
   packet.extract<ethernet_t>(hdr.ethernet);
   transition select(hdr.ethernet.etherType, pvs.is_member(hdr.ethernet.etherType)) {
       (_, true) : parse_mpls;
       (0x8100, _): parser_vlan;
       default: accept;
   }
}

Step 2. lower the value set to the first proposal in bmv2 backend.
ValueSet<bit<16>>(4) pvs;
state parse_ethernet {
   packet.extract<ethernet_t>(hdr.ethernet);
   transition select(hdr.ethernet.etherType) {
       pvs: parse_mpls;
       0x8100: parser_vlan;
       default: accept;
   }
}

Step 3. generate JSON
```

However, the above approach does not fully address the problem, because, in bmv2 backend, we still need to ensure the problem will type check [see here](https://github.com/p4lang/p4c/blob/master/backends/bmv2/backend.cpp#L245) before generating the JSON configuration. 

So it seems that, no matter which proposal we accept, the solution will involve somehow change the type checker to accept the name of `pvs` as the label to the select case, which may not be a compile-time constant, and may need a 'hack' to figure out the type of the values in the value_set.

Not sure how to proceed with the implementation, any suggestion is welcome.

**Comment by vgurevich** [comment]
I definitely like the first option better from the clarity point of view. However, I am also a little worried that the width of the input to the parser value set (16 bits of Ethertype) is implicitly defined by its usage and not in the declaration. I know, this is how P4_14 does it, but I think we should fix that too.

**Comment by jafingerhut** [comment]
Out of curiosity, do you have a snippet of BMv2 JSON that simple_switch accepts today that you believe is a correct way to represent a parser value set in BMv2 JSON, and a corresponding P4_14 or P4_16 source program that it was compiled from, and could attach them both to a comment here (or link to them if they are already in a repo or comment somewhere)?

I was wondering whether having a different way of representing parser value sets in the BMv2 JSON file might make the currently-defined PSA spec easier to implement in the compiler.  That approach has its own issues, I realize, in requiring a different implementation in simple_switch.  I'm just semi-naively asking questions now, to see if any useful ideas turn up.

**Comment by hanw** [comment]
Here are the source and output for [issue946.p4 ](https://github.com/p4lang/p4c/blob/master/testdata/p4_14_samples/issue946.p4)

[issue946.zip](https://github.com/p4lang/p4-spec/files/1680526/issue946.zip)

**Comment by hanw** [comment]
Another slight more complex example.
[issue946-1.zip](https://github.com/p4lang/p4-spec/files/1680536/issue946-1.zip)

**Comment by mbudiu-vmw** [comment]
We should separate as much as possible the problem of specifying value sets from implementing them.
Moreover, the fact that we can implement them in BMv2 does not mean that they will be supported efficiently by another back-end where efficiency is paramount.

We should also focus on the most important scenarios - these are the ones we have to get right.

For example, I believe that using TWO different value sets in a single select statement is not a core scenario; in fact, the way I imagine these are compiler into a ternary table, the system will only be able to handle one table per parse state.

Here is how I can see this done using the `is_member` method:

1) Programs will tend to look like this:
```
    state start {
        packet.extract(hdr.ethernet);
        transition select(pvs.is_member(hdr.ethernet.etherType)) {
            true : parse_mpls;
            default: accept;
        }
    }
```
I.e., there is at most one call to is_member for each select. Even today the front-end will lower this to the following program:

```
    state start {
        packet.extract<Ethernet>(hdr.ethernet);
        tmp_0 = pvs.is_member(hdr.ethernet.etherType);
        transition select(tmp_0) {
            true: parse_mpls;
            default: accept;
        }
    }
```

Notice that the method call has been pulled outside of the select statement, because the compiler does not want the select expression to have possible side-effects.

Now the only thing that needs to be done in the back-end is to compile the assignment statement of tmp_0 into a select statement that uses a value_set. That select statement always has exactly two branches, for true and false, for the is_member method call.

I think that targeting BMv2 will thus be easy.

What we have to worry about is actually generating very efficient code for another target. That will require more work.

**Comment by mbudiu-vmw** [comment]
To summarize, the BMv2 back-end will only have to consider one code pattern: 
```
bool temporary = value_set.is_member(arguments)
```
The front-end does already everything else needed.

**Comment by vgurevich** [comment]
@mbudiu-vmw ,

Actually, using multiple parser value sets in the single select statement (one for each transition) is a very typical scenario.

**Comment by mbudiu-vmw** [comment]
@vgurevich All my statements apply to that case as well; the compiler will pull out each method invocation in a separate assignment. Another, more important question, is how you would compile such a case in hardware. Is it one table with a product of tuples, or two sequential tables?

**Comment by vgurevich** [comment]
@mbudiu-vmw ,

In hardware it s effectively a ternary match table with some entries pre-filled from the program and some that are going to be filled by the control plane at runtime. What the extern will really do is reserve the appropriate number of entries in that TCAM and at the same time provide the handle (name) for the control plane APIs so that they know what to fill.

**Comment by mbudiu-vmw** [comment]
This is excellent. This is where the discussion should start: externs were added into P4-16 exactly to model capabilities of the hardware that have efficient hardware accelerators. So we should forget about how value_sets were defined in P4-14, and we should talk about what the best model in P4-16 would be.

To me it looks like this is very well modeled by the following abstract:

```
extern ValueSet<K, V> {
    ValueSet();
    V lookup(in K key);
}
...
parser p(...) {
    ValueSet<tuple<bit<32>, bit<16>>, bit<2>> dispatch;
    state some {
        transition select (dispatch.lookup({ hdr.h1.f1, hdr.h2.h3 })) {
              0b00: some_state;
              0b01: some_other_state;
              0b10: third_option;
              default: reject;
        }
    }
}
```

To me this looks like the best model for a value_set as you describe it: it maps a tuple to the actual set you care about. This should also be very easy to compile to your target.

I don't know yet if it is easy to compile to bmv2, but that should be a secondary consideration.

**Comment by hanw** [comment]
We had a discussion on the value set in the PSA meeting today. Here is a write-up on my thoughts on value-set and what it means:

The hardware implementation of a parser on a high-performance device uses a ternary match table and action memory to implement a state machine. The implementation shares a lot of common functionalities with the match-action in the control block, with the exception that the parser supports loop, which makes it the only turing-complete part of the P4 language. 

If we take a different view of the P4 language and divide the language into two parts: a core part that with supports `table`, `action`, and another sub-language for describing the state machine in a parser (`parser`, `select`, `transition`...). Further, if we imagine the sub-language for parser can be lowered to the core language with only `table` and `action`. Then, the following program 

```
parser myParser (packet_in buffer, out header_t header) {
    state start {
        transition parse_ethernet;
    }

    state parse_ethernet {
        buffer.extract(header.ethernet);
        transition select(header.ethernet.etherType) {
            // value_set : parse_trill;
            0x8100 : parse_vlan_tag;
            0x0800 : parse_ipv4;
        }
    }

    state parse_ipv4 {
        buffer.extract(header.ipv4);
        transition select(header.ipv4.fragOffset, header.ipv4.protocol) {
            6 : parse_tcp;
            17 : parse_udp;
        }
    }
}
```

can be rewritten into the core language as below:

```
// Q: what is the best way to express value_set to the above example?

// To answer the question, let us look at what happens if we implement
// parser using only the table and action constructs. The goal is to
// lower the parser sub-language to a core language that is much
// closer to hardware implementation and understand what exactly is
// value set in hardware.

enum State_t {
  ETHERNET,
  VLAN,
  IPV4
}

struct ethernet_t {
  bit<48> srcAddr;
  bit<48> dstAddr;
  bit<16> etherType;
}

struct header_t {
  ethernet_t ethernet;
}

module myParser (packet_in buffer, out header_t header) {
    bit<8> next_state;

    action set_next_state(in bit<8> state) {
        next_state = state;
    }

    action extract_ethernet() {
        header.ethernet = buffer.extract<ethernet_t>();
        set_next_state(State_t.ETHERNET);
    }

    table start {
        key = { };
        actions = { extract_ethernet; };
        const entry = { * : extract_ethernet(); }
    }

    action handle_vlan_tag() {
        header.vlan = buffer.extract<vlan_t>();
        set_next_state(State_t.VLAN);
    }

    action handle_ipv4() {
        header.ipv4 = buffer.extract<ipv4_t>();
        set_next_state(State_t.IPV4);
    }

    table parse_ethernet {
        key = {
            hdr.ethernet.etherType : exact;
        };
        actions = { handle_vlan_tag; handle_ipv4; };
        // turns out parser compile-time constant select case are
        // corresponding to the static entries in a table.
        const entry = {
            0x8100 : handle_vlan_tag();
            0x0800 : handle_ipv4();
        }

        // the observation is 'valueset', which is probably a
        // misnomer, corresponding to the DYNAMIC entries in a table.
        // and it is completely control-plane concept. The only
        // implication of 'valueset' on dataplane is the size of the
        // table.
        size = 2 + sizeof(valueset);
    }

    action handle_tcp() {
        header.tcp = buffer.extract<tcp_t>();
        set_next_state(State_t.TCP);
    }

    action handle_udp() {
        header.udp = buffer.extract<udp_t>();
        set_next_state(State.t.UDP);
    }

    table parse_ipv4 {
        key = {
            ipv4.fragOffset : exact;
            ipv4.protocol   : exact;
        };
        actions = { handle_tcp; handle_udp; };
        const entry = {
            6 : handle_tcp();
            17 : handle_udp();
        }
    }

    while(next_state != State_t.END) {
        if (next_state == State_t.START) {
            start.apply();
        } else if (next_state == State_t.ETHERNET) {
            parse_ethernet.apply();
        } else if (next_state == State_t.IPV4) {
            parse_ipv4.apply();
        } else if (next_state == State_t.VLAN) {
            parse_vlan.apply();
        }
    }
}
```

The exercise is NOT to propose changing how we express parser in P4. In contrary, the language we have in P4-16 to express parser is quite nice and succinct. The purpose of the exercise is to shed some lights on what value-set really is.

If you take the view that both parser and control can be expressed with a common match and action abstraction, and in P4-16 we provide the parser sub-language as a convenient sugar. Then the main difference between the parser and control block (aside from the fact that parser can have a loop) is: In control, the common case is that the entries in the table are dynamically inserted from control plane, and less common case is the entries are specified in P4 using the `const entries` attributes. On the other hand, in parser, the common case is the entries are static and equivalent to the `const entries` in P4 table, and the less common case is when the entries are dynamically programmed by the control plane. The less common case **is** value_set.

Now, we have P4 syntax to express both dynamic and static entries in control block tables. And we have a P4 syntax to express the static entries in parser (compile-time constant for select case). But we do not have a P4 syntax to express the dynamic entries in parser, i.e., parser value set.

Now here is the question: if both dynamic and static entries for control block table have native support in the P4-16 language, why would we not have native support for the dynamic entries in parser? Why do we need to resort to the `extern` mechanism to express a concept that is part of core construct in P4-16? The `extern` mechanism is really meant to express hardware building block that is not part of the core P4-16 constructs (parser, control).

I think the fundamental problem is the parser sub-language is missing the syntax to express the dynamic aspect of a parser construct, and a proper solution is to extend the parser sub-language to add in that expressive power. On the contrary, using an `extern` for value-set is not a proper solution and it will be a hack no matter what we propose.

**Comment by mbudiu-vmw** [comment]
Not all targets have the same capabilities. For example, many software targets do not have TCAMs and thus ternary matching. P4-16 moved into extern constructs capabilities which are deemed to be particular to only a specific class of targets. 

I can flip the tables on you and claim that many software targets can run loops, and to do deep packet inspection we have to add loops to P4.

Something should be part of the core language if it is broadly applicable to many classes of targets. For specific targets we should use externs - at least as long as we can express what we need with externs in a concise way. 

Take a look at my previous proposal and tell me where it does not fit the capabilities desired. It looks very simple to me.

**Comment by hanw** [comment]
This example does not capture dynamic entries in parser state. (I will try not use the name value-set, as I think the name does not capture what the thing is)

```
extern ValueSet<K, V> {
    ValueSet();
    V lookup(in K key);
}
...
parser p(...) {
    ValueSet<tuple<bit<32>, bit<16>>, bit<2>>() dispatch;
    state some {
        transition select (dispatch.lookup({ hdr.h1.f1, hdr.h2.h3 })) {
              0b00: some_state;
              0b01: some_other_state;
              0b10: third_option;
              default: reject;
        }
    }
}
```

A dynamic entry in a parser state machine means the entries are not known at compile time. The entry in programmed at runtime by the control plane. The above example assumes 0b00, 0b01, 0b10, which are all static, compile-time constant entries.

**Comment by mbudiu-vmw** [comment]
The entries are programmed by the control-plane into the value_set itself - which exposes a table API to the control-plane. The dispatch method returns the index of the label to jump to for a specific key.
Think of 0b00 being "all keys that map to target 00".

**Comment by vgurevich** [comment]
@mbudiu-bfn ,

I understand the input to the extern, but I do not quite understand the output. It seems to be fairly similar with the previous proposal, where a parser value set would output a boolean to indicate, whether the input matched its contents or not. What do these two bits mean? 

I suggest that we think about P4_14 approach in a slightly different way. In the construct:
```
parser_value_set tpids;
parser_value_set trill_etype;
parser_value_set encap_only_etype;

parser parse_ethernet {
    extract(ethernet);
    return select(ethernet.etherType) {
        tpids: parse_vlan_tag;
       0x800: parse_ipv4;
       0x806: parse_arp;
       trill_etype: parse_triil;
       0x86dd: parse_ipv6;
       default: ingress;
       encap_only_etype: parse_my_special_encap_only_header;
    }
}
```
one way to think of the parser value sets as something that returns a list of cubes (because that's what you can specify as a constant on the left of the colon). But another way to think of these parser value sets is labels, that basically signify the place in the program machine code (be it TCAM entries or whatever) that the control plane can alter using some specialized APIs (that in case of parser value set do somewhat resemble regular table APIs, but are somewhat different too).

I do believe that there is a lot of other uses for the externs that simply give a name to a place in the program that can be changed by the control plane. In some cases they can return a clear type, e.g. you can have "constants" that are actually run-time parameters
```
extern constant_extern timeout;

control bla {
    apply {
        . . . 
       if (m.current_tstamp - m.previous_stamp > timeout.get()) {
      . . .
      }
}
```
Fixing the value inside the if() statement code is much more efficient than getting it out pf the table and then carrying in the metadata.

Another, wild idea is that you can have an extern, that represents a control and what that means is that the code for that control can be written by the SW at run-time (e.g. by pushing some P4 code in it) and so on.

So, in other words, can we think of special class of externs that can specify the point where the control plane can "flex" the program, that is otherwise static?

**Comment by mbudiu-vmw** [comment]
Yes, Vladimir, this is exactly what I am saying.

Assuming that enums are serializable, you can rewrite my previous example as follows:

```
enum NextState {
   SomeState,
   SomeOtherState,
   ThirdOption
}

parser p(...) {
    ValueSet<tuple<bit<32>, bit<16>>, NextState>() dispatch;
    state some {
        transition select (dispatch.lookup({ hdr.h1.f1, hdr.h2.h3 })) {
              NextState.SomeState: some_state;
              NextState.SomeOtherState: some_other_state;
              NextState.ThirdOption: third_option;
              default: reject;
        }
    }
}```

The values I had used 0b00, 0b01 were just explicit encodings of an enum - because today there is no way to send enums from the control plane.

This is exactly one of the important uses for externs: they are an API in the dataplane where the dataplane and the control-plane can communicate.

**Comment by vgurevich** [comment]
@mbudiu-bfn ,

This solution can work, but I am afraid it will not be practical unless there is a good syntax for compile-time initialization of this extern. We can certainly think of the current select statement as such an initializer, but note that it looks at the key (e.g. Ethertype) and not the output (i.e. the state enum). 

In general, I find that using tables in P4 to hard-code subtle algorithms makes the programs very difficult to understand without looking at the control plane that populates these tables. Table initializers in P4_16 definitely help, but even then the program is not easy to read (not to mention the fact that currently these tables become const). So, I would try to find alternative solutions that use easy-to-read syntax to express such ideas.

**Comment by hanw** [comment]
@mbudiu-bfn, your most recent proposal is a little different from the P4-14 parser value set. The way P4-14 value set is specified, each set has a single next state. The new proposal allows each set to have multiple next state transitions.

**Comment by mbudiu-vmw** [comment]
As I said, I think we should focus on the problem that we need to solve and not on the P4-14 design.
Also, many states is a generalization of one state - so you should be able to compile P4-14 programs using this design. To get one transition just instantiate a `bit<1>` for V.

**Comment by mbudiu-vmw** [comment]
Compile-time initialization is a new requirement, which I have not heard before.
Since ValueSet is really very much like a table you should use a syntax like table initializers, but where you replace the action with a value of type V. This is not possible with the current P4-16 grammar, but it would not be a big extension. 
We also have a syntax for extern initializers as an experimental extension to set up the abstract methods. We would need to fit this initializer into that somehow.

But P4-14 does not have initializers, so maybe we should focus on a design without compile-time intializers first.

**Comment by mbudiu-vmw** [comment]
To me the first and most important step is to understand whether the Key-Value map API is suitable:
1) it does what you need
2) is compilable for an ASIC
3) subsumes P4-14 functionality
4) A secondary aspect is how to generate code for bmv2, but I am pretty sure that can be done too, even if it is not in the most efficient way.

**Comment by mbudiu-vmw** [comment]
Are you guys clear about how my proposal would work?
If not maybe I can write it up in somewhat more detail.
We can also try to have a brief conference call to discuss it live.

**Comment by vgurevich** [comment]
@mbudiu-bfn ,

I think a conference call or an in-person meeting is the right thing to do. When are you available? Want to stop by for lunch? As the clarification, I would like to see the example I've previously written in P4_14 re-coded in P4_16:
```
@pragma parser_value_set_size 4
parser_value_set tpids;
parser_value_set trill_etype;
parser_value_set encap_only_etype;

parser parse_ethernet {
    extract(ethernet);
    return select(ethernet.etherType) {
        tpids: parse_vlan_tag;
       0x800: parse_ipv4;
       0x806: parse_arp;
       trill_etype: parse_triil;
       0x86dd: parse_ipv6;
       default: ingress;
       encap_only_etype: parse_my_special_encap_only_header;
    }
}
```

**Comment by mbudiu-vmw** [comment]
There is a P4 design meeting scheduled for next Monday, but there was no announcement yet. That would be a good chance to meet.

Here is how to model this:

```
enum Dispatch { tpids, trill_etype, encap_only_etype }

parser p(...) {
    ValueSet<bit<16>, Dispatch>(4) etherDispatch;
    state parse_ethernet {
        packet.extract(hdr.ethernet);
        Dispatch d = etherDispatch(hdr.ethernet);
        transition select (hdr.ethernet, d) {
              (_, Dispatch.tpids): parse_vlan_tag;
              (0x800, _): parse_ipv4;
              (0x806, _): parse_arp;
              (_, Dispatch.trill_etyle): parse_trill;
              (0x86dd, _): parse_ipv6;
              (_, Dispatch.encap_only_etype): parse_my_special_encap_only_header;
              default: accept;
}}}
```

An alternative encoding puts more burden on the initializer, but less on the compiler, and indeed would require proper initialization by the control-plane for mapping 0x800 to Dispatch.ipv4, etc.

```
enum Dispatch { tpids, trill_etype, encap_only_etype, ipv4, ipv6, arp }

parser p(...) {
    ValueSet<bit<16>, Dispatch>(4) etherDispatch;
    state parse_ethernet {
        packet.extract(hdr.ethernet);
        Dispatch d = etherDispatch(hdr.ethernet);
        transition select (d) {
              Dispatch.tpids: parse_vlan_tag;
              Dispatch,ipv4: parse_ipv4;
              Dispatch.arp: parse_arp;
              Dispatch.trill_etype: parse_trill;
              Dispatch.ipv6: parse_ipv6;
              Dispatch.encap_only_etype: parse_my_special_encap_only_header;
              default: accept;
}}}
```

I agree that the P4-14 syntax is more natural.

**Comment by mbudiu-vmw** [comment]
The nice thing about the second syntax is that it would probably be trivial to generate code for an architecture like the one you describe. We can argue that the compiler may be able to derive the second representation from the first one.

**Comment by mbudiu-vmw** [comment]
If you know that these labels are disjoint (which your program does not guarantee), there is a third way you can write this, which is much cleaner, and arguable as easy for the compiler to use to convert to the second form:

```
enum Dispatch { tpids, trill_etype, encap_only_etype }

parser p(...) {
    ValueSet<bit<16>, Dispatch>(4) etherDispatch;
    state parse_ethernet {
        packet.extract(hdr.ethernet);
        transition select (hdr.ethernet) {
              0x800: parse_ipv4;
              0x806: parse_arp;
              0x86dd: parse_ipv6;           
              default: parse_ethernet_dynamic;
    }}
    state parse_ethernet_dynamic {
         Dispatch d = etherDispatch(hdr.ethernet);
         select (d) {
              Dispatch.tpids: parse_vlan_tag;
              Dispatch.trill_etyle: parse_trill;
              Dispatch.encap_only_etype: parse_my_special_encap_only_header;
              default: accept;
}}}

```

**Comment by vgurevich** [comment]
@mbudiu-vmw ,

The third option (using separate states for static and dynamic dispatching) has been previously discussed, since it was the only viable one for the previous (.is_member()) proposal. I still consider it more like a workaround and I have serious doubts that the compilers will be able to implement it efficiently by merging the states properly.

The second option has a major flaw, since it does not expose the actual values of the Ethertypes for all "static" cases, moving them somewhere into the control plane. In my view, this is a no-go.

The first option is kind of OK, but it does not model the P4_14 code properly, since it merges 3 independent parser value sets (one with 4 elements and two others with some implementation-defined number of elements >=1) into one object. This object will be much more difficult to control: for example in P4_14 case you do not need to specify entry priorities (since the outcome is the same), but here you will need to and so on. If you attempt to use `tpids` parser value set in another place (e.g. in parse_vlan_tag state) it will further complicate things.

So, the only thing I can totally agree with you on so far is your last statement on the superiority of P4_14 syntax. :)

**Comment by vgurevich** [comment]
@mbudiu-bfn ,

Actually, the comment about your first example merging three independent objects into one applies to all the examples.

**Comment by mbudiu-vmw** [comment]
You forget that in the implementation you actually have a single object - the table whose entries you need to populate. The third version is very close to the hardware you described, and it would be trivial to support in a real back-end.

**Comment by vgurevich** [comment]
@mbudiu-vmw,

The fact that sometimes it is a single TCAM should be totally irrelevant. This is not how P4 programmers think about the problem. Moreover, since in TCAM implementations (such as kangaroo parsers) there is is a mandatory match field called "state", I would argue that it effectively splits one big TCAM into a number of virtual ternary tables of variable lengths (one per state). 

We are trying to develop a language for people who write data plane algorithms and we must offer familiar objects. Trying to directly represent the HW is the last resort option.

**Comment by hanw** [comment]
https://github.com/p4lang/p4c/pull/1158 implemented an experimental feature to support `value_set` as a first class type that behaves like a set and is only supposed to be declared in parser block. The sizing of `value_set` is done via annotation `@size` or `@parser_value_set_size` (exact keyword TBD).

The rationale behind why introducing this first class type in P4-16 is explained in this thread.

```
value_set<T> pvs;
```

The name of a parser value set can be used as the select case in a parser state, just as we had in P4-14.

```
value_set<bit<16>> pvs0_0;
value_set<bit<16>> pvs1_0;
state parse_ethernet {
   packet.extract<ethernet_t>(hdr.ethernet);
   transition select(hdr.ethernet.etherType) {
       pvs0_0: accept;
       pvs1_0: parse_inner_ethernet;
       default: accept;
   }
}
```

**Comment by jafingerhut** [comment]
This PR proposes restrictions on how parser value sets may be used in a P4_16 program.  The restrictions are very similar to the restrictions on the use of tables, which parser value sets as described in the recent comments are very similar to: https://github.com/p4lang/p4-spec/pull/586

That PR _does not_ currently describe what parser value sets _are_.  Such a PR should be created before merging in #586.

**Comment by vgurevich** [comment]
If value_set is a new first class object, then why not introduce a distinct syntax that would allow size specification, instead of trying to use annotations? I mean, we now have 3 classes of stateful objects (tables, externs and value_sets) and each uses a different way to specify such a fundamental property as size: tables use a property, for externs it is an instantiation parameter and for value_sets we want to use an annotation?

**Comment by mbudiu-vmw** [comment]
What is the "size" of a value set?

**Comment by hanw** [comment]
The 'size' of a value set is the maximum number of elements (e.g., entries in the parser tcam) that can be inserted by the control plane into the set.

**Comment by mbudiu-vmw** [comment]
Is this enough information for a back-end to allocate resources?
Also, what if you have a dynamically-allocated value_set and you don't want to specify a maximum size?

**Comment by vgurevich** [comment]
@mbudiu-vmw ,

This is precisely the same problem as with tables. In P4_14 there were attributes min_size, max_size and size to take of this and I think that was quite reasonable design. For tables in P4_16 we agreed to standardize the "size" attribute, but implementations that support dynamic sizing can still define their "min_size" or "max_size" or whatever else they need. I think value_sets need something similar, but I think it should be done as a language feature and not as an annotation, since we are defining a new, first-class object type.

**Comment by mbudiu-vmw** [comment]
You are right, it's the same as for tables. But so far the size attribute has not been standardized, exactly because some architectures may not need it, and it's meaning is not clear.

**Comment by hanw** [comment]
Status update:
I propose the following changes to IR to support `size` attribute as a constructor parameter to value_set. The previous proposal described in this PR to support value_set as if it is a variable with `set` type will be superseded by the new IR modification.

```
 parserLocalElement
     : constantDeclaration             { $$ = $1; }
+    | valueSetDeclaration             { $$ = $1; }
     | instantiation                   { $$ = $1; }
     | variableDeclaration             { $$ = $1; }
     ;

/************************* VALUE_SET *****************************/
valueSetDeclaration
    : optAnnotations
        VALUESET "<" typeArgumentList ">" optConstructorParameters name ";"
        { auto pl = new IR::ParameterList(@6, *$6);
          $$ = new IR::P4ValueSet(@7, *$7, $1, $4, pl); }
    ;

class P4ValueSet : Declaration, IAnnotated {
   optional Annotations       annotations = Annotations::empty;
    Vector<Type>                 typeArguments;
    ParameterList                 parameters;
    Annotations getAnnotations() const override { return annotations; }
}
```

Given the above IR definition, a parser value set will syntactically look like:

```
value_set<bit<32>, bit<32>>(4) pvs;
value_set<bit<32>> (8) pvs;
```

We could modify the IR to restrict the type parameter of parser value set to be either a tuple type, a bit type or a type name, which will give us the following:
```
value_set<tuple<bit<32>, bit<32>>> (4) pvs;
value_set<bit<32>> (8) pvs;
```

I prefer the first one, but can accept the second if decided otherwise.
